docs/native-modules-async.md

```diff
@@ -25,94 +25,122 @@ NativeModules.SimpleHttpModule.GetHttpResponse('https://microsoft.github.io/reac
 
-Now we want to expose the method that performs the HTTP request. It is recommended, and it is the default, to write these functions asynchronously. Writing asynchronous code in C# is pretty straight-forward and intuitive with the `async` and `await` keywords and the `Task<T>` types. 
+> **Note:** Using `JSValue` here is a workaround - we actually want to return a `JSValueObject` but using `ReactPromise<JSValueObject>` in C# will cause a runtime crash [#5617](https://github.com/microsoft/react-native-windows/issues/5617). Functionally we can still resolve a `ReactPromise<JSValue>` with a `JSValueObject`.
 
-> If you're not familiar with writing asynchronous C# code, see [Call asynchronous APIs in C# or Visual Basic](https://docs.microsoft.com/en-us/windows/uwp/threading-async/call-asynchronous-apis-in-csharp-or-visual-basic) and [Asynchronous programming](https://docs.microsoft.com/en-us/dotnet/csharp/async) that will teach you the concepts if you are not familiar yet.
+All method parameters before this final promise are the input parameters we expect to be marshaled in from the JS. In this case, we want a single string for the URI to request.
 
-The function signature for a typical web request, annotated with the `[ReactMethod]` attribute will look like:
+The `promise` object is our interface for handling the promise and marshaling a result to the JS. To do so we simply call `promise.Resolve()` with the result object (if the operation was a success) or `promise.Reject()` with an error (if the operation failed).
 
-```cs
-    [ReactMethod]
-    public async Task<string> GetHttpResponseAsync(string uri) {
-      ...
-    }
-```
+Now that we know how to return results, let's prep `GetHttpResponseAsync` to take in a `ReactPromise<JSValue>` parameter and use it:
 
-You are now free to fill in the logic like:
+```csharp
+static async Task GetHttpResponseAsync(string uri, ReactPromise<JSValue> promise)
+{
+  // Create an HttpClient object
+  var httpClient = new HttpClient();
 
-```cs
-      // Create an HttpClient object
-      var httpClient = new HttpClient();
+  // Send the GET request asynchronously
+  var httpResponseMessage = await httpClient.GetAsync(new Uri(uri));
 
-      // Send the GET request asynchronously
-      var httpResponseMessage = await httpClient.GetAsync(new Uri(uri));
+  var statusCode = httpResponseMessage.StatusCode;
+  var content = await httpResponseMessage.Content.ReadAsStringAsync();
 
-      var statusCode = httpResponseMessage.StatusCode;
-      var content = await httpResponseMessage.Content.ReadAsStringAsync();
-      
-      return content;
+  // Build result object
+  var resultObject = new JSValueObject();
+
+  resultObject["statusCode"] = (int)statusCode;
+  resultObject["content"] = content;
+
+  promise.Resolve(resultObject);
+}

```

Comment:
This isn't right... This has a return type of Task but doesn't return anything. @dannyvv did some work to support task-style returns before. Let's make sure to loop him in.